### DISCUSSION TITLE
Does USBHostMID work only with Teensy board?

### DISCUSSION THREAD

**Discussion by josehdx** [tttapa/Control-Surface#794] [Q&A]
Dear Dev

I have some "deep" technical questions. I want to connect my external MIDI controller to a new ESP32 (S3) board (no bought yet)

As far as i know there are some libraries on your contoller that allows to make to board work as Host (USBHostMIDi ) but it looks like only works with Teensy' boards, because, until now, those were mainy the only ones with the USB host capabilities.

I m not an expert but the new ESP32-S3 boards has the same USB Host capabilities, but i assume your library needs to be updated in order to make them work with the ESP32., am i right?

Another question regarding compatibility, i have been talking about the new S3 versions, which looks like an upgrade of the old ESP32, so, puting aside the MIDI USB host, does the s3 boards work as expected with your library as well or is there any compatibility issues?

My main goal is, if the library works, to connect the floor MIDI controler via USB to the new S3 board, and besides powering the floor controler with the battery of the S3, i want to make the s3 as a bridge  transmiting the MIDI coming  messages from the floor device and send it via BLE. I think it should bepossible as there are some examples on Vguitar forums (https://www.vguitarforums.com/smf/index.php?topic=25185.msg183442#msg183442) (well, is what one guy mentioned over there)

So, do you think is possible, or is there any blocker with the library?

**Comment by tttapa** [comment]
> I m not an expert but the new ESP32-S3 boards has the same USB Host capabilities, but i assume your library needs to be updated in order to make them work with the ESP32., am i right?

No changes should be necessary to the library. If you find a USB Host library for your specific chip, you can plug it into the existing MIDI USB machinery of the Control Surface library, see e.g. https://tttapa.github.io/Control-Surface-doc/Doxygen/d7/d5f/Custom-USB-MIDI-Backend_8ino-example.html and the links therein.

> does the s3 boards work as expected with your library as well or is there any compatibility issues?

I don't own an ESP32-S3 so I've never tested it.

> So, do you think is possible, or is there any blocker with the library?

At first sight, this seems feasible.

**Reply by josehdx** [reply]
I think i have found the libraries for the USB on the S3, too much information to be understood first :) i will try to keep and update once that my new board arrives, but yeah, it looks possible to do. Thanks

**Reply by pukulsesuatu** [reply]
> I think i have found the libraries for the USB on the S3, too much information to be understood first :) i will try to keep and update once that my new board arrives, but yeah, it looks possible to do. Thanks

nice project @josehdx , waiting for the result

**Reply by josehdx** [reply]
> No changes should be necessary to the library. If you find a USB Host library for your specific chip, you can plug it into the existing MIDI USB machinery of the Control Surface library, see e.g. https://tttapa.github.io/Control-Surface-doc/Doxygen/d7/d5f/Custom-USB-MIDI-Backend_8ino-example.html and the links therein.
> 
> > does the s3 boards work as expected with your library as well or is there any compatibility issues?
> 
> I don't own an ESP32-S3 so I've never tested it.
> 
> > So, do you think is possible, or is there any blocker with the library?
> 
> At first sight, this seems feasible.

I have found the required host file from the ESP IDF libraries, nevertheless im not 100% sure if this file from IDF library is compatible with the surface library. I will test it in the weekend, i just want to share the link now.

https://docs.espressif.com/projects/esp-idf/en/v5.0/esp32s3/api-reference/peripherals/usb_host.html#examples

"The API of the USB Host Library is separated into the following header files. However, it is sufficient for applications to only #include "usb/usb_host.h" and all of USB Host Library headers will also be included"

**Reply by Gershy13** [reply]
Hey, just wondering if you got this to work? I've got an esp32-s2 which also has usb support. I'm trying to build a midi controller. Did you manage to get it to work? If so, how?

**Reply by Trickster-git** [reply]
No, it works on Pi Pico and atmega 32u4.
I have esp32-s3 i can`t run it as usb midi device.

**Reply by Gershy13** [reply]
Did the library/header file that @josehdx linked not work?

**Reply by Trickster-git** [reply]
My be it work but, it very powerful Kungfu, understand how use Custom-USB-MIDI-Backend. For me easy way, buy more new boards and try it.

**Reply by josehdx** [reply]
> Did the library/header file that @josehdx linked not work?

Hi. Im not an expert on the topic, but looks like the link I shared for usb host works only under espressif idf. This means it should not be compatible with Arduino, hence with thi surface library, but i might be wrong. 

Still there are another under progress libraries that might work, but the same developer mentioned that we should have some deep knowledge about how midi over usb works.

https://github.com/chegewara/EspTinyUSB

https://github.com/chegewara/esp32-usb-v2

For my wireless (ble converter) project, i think i will stick with the current usb connectivity that my midi device has. And actually, if i can find a way, i will probably use the cheguewara libraries to set the s3 board as usb midi device for other projects, but i will leave it on pending now

**Reply by Gershy13** [reply]
Adafruit TinyUSB is working for me with the ESP32-S2 and the default Arduino MIDI library, have not tested it with control surface yet.

**Reply by josehdx** [reply]
> Adafruit TinyUSB is working for me with the ESP32-S2 and the default Arduino MIDI library, have not tested it with control surface yet.

Are you using Arduino ide or vscode/platformio? Then i would say only the backend should be set but haven't tried this. Actually i need to order the usb c breakout first😅

**Reply by Gershy13** [reply]
> > Adafruit TinyUSB is working for me with the ESP32-S2 and the default Arduino MIDI library, have not tested it with control surface yet.
> 
> Are you using Arduino ide or vscode/platformio? Then i would say only the backend should be set but haven't tried this. Actually i need to order the usb c breakout first😅

I am using Arduino IDE. No clue how to use the others lol. My board (ESp32-S2-DevkitC) has two usb ports on it (one UART, one USB)
